feat: add `extract` as a new mode

kopecs opened the pull request:
This PR adds an additional mode `extract` which allows the extraction of source code in a different language from the containing file. Namely, an extract rule with `languages: [ X ]` and `dest-language: Y` will run the pattern against files of language X and produce new targets in language Y which search & taint rules will be run against. Currently extract rules will not run against the result of extract rules.

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)


Test plan:

You can manually test the changes here with this small example:

`test.html`:
```html
<!DOCTYPE html>
<html>
    <body>
        <h2>Demo JavaScript in Body</h2>

        <p id="demo">A Paragraph</p>

        <button type="button" onclick="myFunction()">Try it</button>

        <script>
            function myFunction() { document.getElementById("demo").innerHTML = "Paragraph changed."; }
        </script>

    </body>
</html>
```

`rules.yaml`:
```
rules:
  - id: extract-javascript-from-html
    mode: extract
    pattern: |
        <script>$...BODY</script>
    extract: $...BODY
    languages:
      - html
    paths:
      include:
        - '*.html'
    dest-language: javascript
  - id: get-elem-id
    mode: search
    pattern: document.getElementById(...)
    languages:
      - javascript
    message: "Found match"
    severity: INFO
```

then the command `semgrep --config rules.yaml test.html` should produce:

```
Scanning 1 file.

Findings:

  test.html
     get-elem-id
        Found match

         11┆ function myFunction() { document.getElementById("demo").innerHTML = "Paragraph changed."; }

Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.

Ran 2 rules on 1 file: 1 finding.
```

Commit:
Be more permissive of whitespace for xml ellipsis

In certain contexts we might have leading/trailing whitespace for an
ellipsis in a text or raw text parse node, but we probably ought to
treat this as an ellipsis regardless.

A better solution would probably be to generally migrate from having
ellipsis be an identifer to having it be a special variant.

Commit:
Add `extract` mode for rules

Commit:
Add `-pp` flag to core to pass extract rule file

It's unclear to me if this should really be a separate flag or not,
given that I'm implementing it using a mode. This introduces a bit of an
artificial barrier and will be a little bit less ergonomic when it comes
to making cli changes later, but is fine for testing and we can perhaps
change to sticking with just using multiple invocations of `-rules` later.

Commit:
Derive show for many more types

Primarily this will help when printing values out to debug

Commit:
Add extracting module

Handles the logic for getting the extracted ranges from a source text
given an extract rule and determining how to map matches in the
extracted portion back to matches for the original text.

Currently just handles each extracted portion as a separate document.

Commit:
Implement logic for running extract rules

Commit:
Update cli progress for core-generated targets

aryx commented on cli/src/semgrep/core_runner.py:
```diff
@@ -126,6 +126,7 @@ class StreamingSemgrepCore:
 
     This behavior is assumed to be that semgrep-core:
     - prints a "." on a newline for every file it finishes scanning
+    - prints a number on a newline for any extra targets produced during a scan
```
Why those things are part of this diff?
Did you change things for the extractor or it's just you use the opportunity to better document things?

aryx replied:
What is an extra target? How semgrep-core can return extra target?

kopecs replied:
This corresponds to a change in `Run_semgrep.ml`: https://github.com/returntocorp/semgrep/blob/524c85af3e8aeb866d80fabcbb0af09ab8fe2daa/semgrep-core/src/runner/Run_semgrep.ml#L520-L522

An "extra target" in this context would be the temporary files containing extracted text. If we want these to get a entry in the progress bar on the python side, we need some way to communicate the number of targets back to python. Right now the cli just calculates it ahead of calling core, so this gives us a mechanism for that.

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -196,6 +196,18 @@ type taint_spec = {
+               "separate"?
+   * combine_extract: extract_mode;
+   *)
+  (* TODO: really want Lang.t | Generic *)
```
I don't see a use for using generic as a dst_lang but who knows.
Good to keep as a TODO for now.

kopecs replied:
I think the use-case would be for things like django templates, where you have a DSL which can appear in specific parts of the file only (although using generic for the whole file might be fine for those cases).

aryx commented on semgrep-core/src/engine/Match_rules.ml:
```diff
@@ -118,7 +118,12 @@ let check ~match_hook ~timeout ~timeout_threshold default_config rules xtarget =
-                           res)
+                           res
+                       | `Extract extract_spec ->
+                           (* TODO: need another module for extract? *)
```
not unless things get really different between search and extract. Anyway your Extract.ml seems good enough for now to try to isolate all extractor specific stuff in a separate module.

kopecs replied:
Okay, sounds good.

aryx commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -0,0 +1,142 @@
+module In = Input_to_core_j
```
A toplevel copyright and a toplevel Prelude section explaining what is the
goal of this module.

aryx commented on semgrep-core/src/extracting/Extract.mli:
```diff
@@ -0,0 +1,22 @@
+(* should only be extract rules *)
```
I guess you can remove the comment now that you enforce Rule.extract_rule below.

Then you can replace by a general comment explaining the goal of the function.

aryx commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -0,0 +1,142 @@
+  (* match combine with
+     | "separate" -> *)
+  res.Report.matches
+  (* want Common.filter_map? *)
```
Common.ml was created a long time ago. If List now has filter_map, better to use it.
There's just Common.map we really prefer over List.map because it's more efficient.

kopecs replied:
I wasn't sure if it made sense to implement the corresponding functionality in a tail recursive way or not. If we don't have it right now would you just rely on the stdlib implementation?

aryx commented on semgrep-core/src/extracting/Extract.mli:
```diff
@@ -0,0 +1,22 @@
+  timeout_threshold:int ->
+  Rule.extract_rule list ->
+  Xtarget.t ->
+  string list ->
```
what is this string list?

kopecs replied:
This is a list of rule ids we're interested in running. We need this information when generating the additional targets as each target has a list of rule ids to run against it. Normally this would be generated for us by the cli, but in this case because we're creating a new input in core we don't have that.

There's probably a type alias I can swap this to, and when I add docs I'll make it more clear.

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -463,28 +463,82 @@ let targets_of_config (config : Runner_config.t)
  * It takes a set of rules and a set of targets and
  * recursively process those targets.
```
maybe add a comment about the extraction step. How does it influence the regular matching process of semgrep_with_rules.

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -463,28 +463,82 @@ let targets_of_config (config : Runner_config.t)
  * It takes a set of rules and a set of targets and
```
maybe you can adjust with "and a set of targets (derived from [config])"

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -463,28 +463,82 @@ let targets_of_config (config : Runner_config.t)
-let semgrep_with_rules config ((rules, invalid_rules), rules_parse_time) =
+let semgrep_with_rules config extractors
+    ((rules, invalid_rules), rules_parse_time) =
   (* Return an exception
```
Maybe you can add before the comment "We want to be tolerant to errors in rules. This is useful
when rules are using advanced features not present in older version of semgrep-core, but
still want this old semgrep-core to process the other (old-compatible) rules".

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -463,28 +463,82 @@ let targets_of_config (config : Runner_config.t)
   | [], err :: _ -> raise (Rule.InvalidRule err)
   | _, err :: _ (* TODO fail fast when only when strict? *) ->
       raise (Rule.InvalidRule err)
   | _ -> ());
```
I guess you can replace the last pattern with | _, [] -> ())

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -463,28 +463,82 @@ let targets_of_config (config : Runner_config.t)
 
+  let rule_ids = Common.map (fun r -> fst r.R.id) rules in
+
+  let extracted_targets, extract_result_map =
```
move this code out of semgrep_with_rules in an helper function.
Maybe add a (* *** *) Extractor helpers section.
semgrep_with_rules() is already long, we don't want to make it too much longer.

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -512,11 +566,14 @@ let semgrep_with_rules config ((rules, invalid_rules), rules_parse_time) =
+    RP.make_final_result file_results rules config.report_time
+      (erules_parse_time +. rules_parse_time)
   in
+  RP.show_final_result res |> pr2;
```
this call to pr2 seems wrong.
use logger if needed.

kopecs replied:
Ah this is a debug print that slipped through, thanks for the catch!

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -546,14 +603,54 @@ let semgrep_with_raw_results_and_exn_handler config =
   let rules_file = replace_named_pipe_by_regular_file rules_file in
+  let extractors_file =
+    if config.extractors_file <> "" then
+      Some (replace_named_pipe_by_regular_file config.extractors_file)
```
I doubt we want to define extractor rule via named pipe. 
I think this replace_named_pipe_... was added so you could do in the shell semgrep  -e foo  $(<echo foo>) or something like that, but for -pp I don't think we need that feature.

kopecs replied:
Hmm actually wouldn't the same logic as in the comment apply, like if you wanted to run with `-pp <(curl $URL)`?

aryx replied:
Oh you re right. Might be good to add (* this allows -pp (curl ...)

aryx commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -546,14 +603,54 @@ let semgrep_with_raw_results_and_exn_handler config =
   try
+    logger#linfo
+      (lazy (spf "Parsing %s:\n%s" rules_file (read_file rules_file)));
+    let timed_extractors =
```
again move in helper function in the 'Extract helper' section to not clutter the original function.

kopecs replied:
Will do.

aryx commented:
Please provide a test plan: section in the PR indicating how to test this PR.
You can add an example of target file, extractor rule, like in tests/OTHER/extractor, and indicate the command to run.

Commit:
Refactor runner code for extract rules

Simplify logic by simply using `-rules` for extract mode rules instead
of having a special flag and performing additional redundant validation.

Break some runner logic out into a helper function.

Commit:
Document `Extract` module; add LGPL header.

Commit:
Disallow extract mode keys for other modes

brendongo commented on cli/src/semgrep/core_runner.py:
```diff
@@ -163,33 +164,51 @@ async def _core_stdout_processor(
         while True:
             # blocking read if buffer doesnt contain any lines or EOF
-            line_bytes = await stream.read(n=bytes_to_read)
+            line_bytes = await stream.readline()
```
I think we went from readlint to bytewise reading because we were filling up the pipe.

brendongo replied:
Relevant: https://github.com/returntocorp/semgrep/issues/4693

I think we could keep track of where we are wrt total and pre-emptively breakout instead of trying to read the last line which is a giant json blob?

kopecs replied:
Issue would be when `read` returns less than `n` bytes; per our conversation I'll just switch the termination condition to being based on the number of files scanned.

aryx commented:
When do you think this will be ready for review and merge? (It's still in draft right now).

Commit:
Fix possible IO issues in CLI

Some updates to the communication between CLI and core could potentially
cause issues like https://github.com/returntocorp/semgrep/issues/4693
due to the amount of bytes we're reading at a time.

Commit:
Don't require message and severity for extract

Currently just emits fake data for these fields. Futuristically these
could be moved to only the modes which require them.

kopecs commented:
Largely good (I can do the tests either with this or in a separate PR), I just want to get the validation working well. The current error messages would be a decently poor regression (imo); it seems like it might make sense to upgrade the version of JSON schema we're using so we can use [`dependentRequired`](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-dependentrequired)---as it stands the validation schema is a bit of a mess.

emjin commented:
Can you include some form of test plan? That can be just pasting the contents of two short files with instructions for what to name them and what command to run. It would be helpful to see just one example of the output, maybe a second to show how bad the error messages become.

kopecs commented:
I'll update the original comment with that. The error messages I'm referring to there are mostly about the json schema validation. You can see the [pytest snapshot diff](https://github.com/returntocorp/semgrep/commit/90278d46f0838a31eff319e3cacb3345d0694c84) for that---basically it just dumps a bunch of keys that don't make sense (cf. behaviour with taint mode keys, as those don't show up).

Commit:
Handle no progess bar case in CLI

Updates the stdout processing to work even when a progress bar is not
produced (namely, when the plan contains only one target).

Commit:
Add severity fallback in CLI

This should be improved, but allows rules whose findings will not be
reported to omit severity and simply have a "dummy" internal value. This
will have some adverse interaction with severity filtering and should be
changed. Perhaps we should have another internal severity for rules like
these?

Commit:
Improve rule schema and update jsonschema

Updates jsonschema from 3.2 to 4.6.1 (types are 4.6.0).

Refactors our jsonschema from being `oneOf` many different mode
properties to instead satisfy `allOf` many implications on modes---this
has the effect of simply assuming the mode of the rule, which improves
diagnostics by avoiding dumping a list of irrelevant or illegal fields.

Commit:
Merge with develop

Resolves python lockfile conflicts

Commit:
Update pipfile for jsonschema transitive deps

Commit:
Update changelog

Commit:
Merge branch 'develop' into cooper/nested-lang-extraction

Commit:
Fix minor schema typos

Commit:
Use draft 7 validator

This avoids the need to update the logic for error message translation
for now.

Commit:
Merge branch 'develop' into cooper/nested-lang-extraction

Commit:
Improve error message for mode-illegal properties

For properties which are allowed generally, but disallowed for a
specific mode, specifically indicate these as disallowed as invalid for
the current mode. There is room for future improvement in these error
messages in the case where a property is only used by one mode, in which
case it should just be a extra property on that specific mode, instead
of a top-level property.

Commit:
Don't tell python about 0 new targets

Commit:
Improve error messages for mutually exclusive keys

Commit:
Update CHANGELOG entry for new format

Commit:
Improve handling for when no plan is used

In the case a plan is not used we do not calculate the number of
targets. This changes the handling around capturing the json output when
the number of targets is not calculated by Python.

github-actions[bot] committed:
Update pytest snapshots

Commit:
Simplify runner logic

Commit:
Correctly allow 1 of [mutex patterns, depends-on]

Previously made depends-on mutually exclusive with all patterns, instead
of requiring at least one of these two groups to appear.

Commit:
Merge branch 'develop' into cooper/nested-lang-extraction

Commit:
Only read exact when we expect exact amounts

... and not when we're potentially reading half a gig.

Commit:
Don't treat numbers in json blob as addt'l targets

Commit:
Add some redundant properties to improve errors

Commit:
Update comment-level documentation

Commit:
Read from file instead of using AST for extract

Commit:
Merge branch 'develop' into cooper/nested-lang-extraction

Resolve test race conditions for dependency-aware

Commit:
Merge branch 'develop' into cooper/nested-lang-extraction

Relock

mmcqd commented on cli/Pipfile:
```diff
@@ -29,5 +29,21 @@ pytest-split = "*"
 pypi-parker = "*"
 pytest-rerunfailures = "*"
 
+# pipenv does not support multiple python versions, but the
+# dependencies of jsonschema differ from < 3.8 to >= 3.8. Therefore we just
+# rely on these directly for the dev environment so that an entry for these is
+# always generated for the lockfile, regardless of the local version of python.
+# If we change to Poetry or another system which allows multiple version of
+# python these could likely be removed.
+#
+# Sourced from
+# <https://github.com/python-jsonschema/jsonschema/blob/afc22f09e74d696ab00be8a711bbc5c2a15327b7/pyproject.toml#L39-L41>
+#
+# These _should not_ be added to setup.py
+importlib_metadata = "*"
+typing_extensions = "*"
+importlib_resources = ">=1.4.0"
```
Will defer to @brendongo on the okayness of this lol

kopecs replied:
@underyx suggested this as a solution to pipenv's lack of support for multiple python versions.

underyx replied:
could you leave a comment explaining the need for this tho?

kopecs replied:
Is the above comment not sufficient? I'm not sure what else specifically to add.

brendongo replied:
Looks sufficient! Feel free to merge and we can add any comment that is missing later if ever

underyx replied:
oh sorry I missed that one; I thought `importlib_resources` was the only line relevant here so I expected the comment directly above this line

the one we have above this block is very well written!

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -0,0 +1,181 @@
+      (erules :> Rule.rules)
+      xtarget
+  in
+  (* mode for combination *)
```
Probably want to delete this or explain what it is

kopecs replied:
I have this fleshed out on the other one.

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -0,0 +1,181 @@
+  res.Report.matches
+  |> List.filter_map (fun m ->
+         match
+           Common.find_some_opt
+             (fun (x, mvar) ->
+               match
+                 Hashtbl.find_opt erule_table m.Pattern_match.rule_id.id
+               with
+               | None -> None
+               | Some r ->
+                   let (`Extract { Rule.extract; _ }) = r.mode in
+                   if x = extract then Some (r, Some mvar) else Some (r, None))
+             m.Pattern_match.env
```
This would be easier to read if it was computed using a helper function. It's hard to patch `match <a bunch of stuff> with ...`

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -0,0 +1,181 @@
+             let* ( (char_offset, line_offset, col_offset),
+                    (start_extract_pos, end_extract_pos) ) =
+               match
+                 Metavariable.mvalue_to_any extract_mvalue
+                 |> Visitor_AST.range_of_any_opt
+                 |> Option.map (fun (start_loc, end_loc) ->
+                        let end_len = String.length end_loc.Parse_info.str in
+                        ( ( start_loc.Parse_info.charpos,
+                            (* subtract 1 because lines are 1-indexed, so the offset is
+                             * one less than the current line *)
+                            start_loc.line - 1,
+                            start_loc.column ),
+                          (start_loc.charpos, end_loc.charpos + end_len) ))
```
Same here. In general I think this function is a little monolithic. Giving names to intermediate results computed by helpers would make it clearer

mmcqd commented on semgrep-core/src/matching/Generic_vs_generic.ml:
```diff
@@ -1359,7 +1359,7 @@ and m_attrs a b =
   match (a, b) with
   (* dots: *)
-  | [ XmlText ("...", _) ], _ -> return ()
+  | [ XmlText (s, _) ], _ when String.trim s = "..." -> return ()
```
What's this for?

kopecs replied:
This is to make ellipses in XML/HTML tags a little easier to work with, so that e.g.,
```
<tag>
  ...
</tag>
```
still hits the above case.

Should I leave this in and add a changelog entry, remove it, or do you have another suggestion?

kopecs replied:
Added changelog entry.

mmcqd commented on semgrep-core/src/runner/Run_semgrep.ml:
```diff
@@ -506,36 +506,87 @@ let targets_of_config (config : Runner_config.t)
+        pr (string_of_int (List.length extract_targets));
+      extract_targets)
+    extract_targets
+  |> Fun.flip
```
I would just use a lambda

mmcqd reviewed (approved):
Basically just style comments, which I know you've largely fixed in another branch. Feel free to just resolve them if they're fixed there.

Commit:
Use a lambda instead of flip

Commit:
Remove resolved todo

Commit:
Add changelog entry for xml change

brendongo commented on cli/setup.py:
```diff
@@ -125,7 +125,7 @@ def find_executable(env_name, exec_name):
     "tqdm~=4.46",
     "packaging~=21.0",
-    "jsonschema~=3.2",
+    "jsonschema~=4.6.1",
```
```suggestion
    "jsonschema~=4.6",
```
So minor version bumps are allowed?

kopecs replied:
Sounds good.

brendongo commented on cli/src/semgrep/core_runner.py:
```diff
@@ -163,33 +166,50 @@ async def _core_stdout_processor(
-        bytes_to_read = 2
+        get_input: Callable[
+            [asyncio.StreamReader], Coroutine[Any, Any, bytes]
+        ] = lambda s: s.readexactly(2)
```
Oooo shiny ✨

Commit:
Allow minor version bumps for jsonschema

Co-authored-by: Brendon Go <brendon.go@gmail.com>

brendongo commented on cli/tests/e2e/snapshots/test_ci/test_github_ci_bad_base_sha/results.txt:
```diff
@@ -132,6 +132,7 @@ Running semgrep-core with command:
 <MASKED>
 <MASKED>
 <MASKED>
+<MASKED>
```
@kopecs did you add any lines that print to stderr? Probably innocuous but might be worth checking what this is exactly https://github.com/returntocorp/semgrep/blob/develop/cli/tests/e2e/test_ci.py#L534

kopecs replied:
Appears to be the extra line from here: https://github.com/returntocorp/semgrep/blob/3395c61a76c77542f2c2238e1c3b001c21459d5c/semgrep-core/src/runner/Run_semgrep.ml#L522-L524

aryx reviewed (commented):
this is great!
